src/nnet3bin/nnet3-compute-batch-from-wavs.cc

```diff
@@ -0,0 +1,205 @@
+// nnet3bin/nnet3-compute-batch-from-wavs.cc
+
+// Copyright 2012-2018   Johns Hopkins University (author: Daniel Povey)
+//           2021        Yan Zhao
+
+// See ../../COPYING for clarification regarding multiple authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+
+#include "base/kaldi-common.h"
+#include "util/common-utils.h"
+#include "base/timer.h"
+#include "nnet3/nnet-utils.h"
+#include "matrix/kaldi-vector.h"
+#include "nnet3/decodable-batch-looped.h"
+#include "online2/online-nnet2-feature-pipeline.h"
+
+using namespace kaldi;
+using namespace kaldi::nnet3;
+typedef kaldi::int32 int32;
+typedef kaldi::int64 int64;
+
+typedef struct _ThreadPara {
+  std::vector<std::string> wavs_file_name;
+  std::vector<std::string> utts;
+  BaseFloatMatrixWriter    *matrix_writer;
+  std::mutex               mtx;
+  int                      index;
+  
+  OnlineNnet2FeaturePipelineInfo *pipeline_info;
+  NnetBatchLoopedComputer        *computer;
+  TransitionModel                *trans_model;
+
+} ThreadPara; 
+
+
+void *ThreadFunction(void *para) {
+  ThreadPara *thread_para = reinterpret_cast<ThreadPara *>(para);
+  int output_dim = thread_para->computer->GetInfo().output_dim;
+  static int package_size = 1024;
+  char package[package_size];
+  FILE *wav_file;
+  int read_len, num_frames_read;
+  
+  while (true) {
+    std::string utt, wav_file_name;
+    
+    {
+      std::unique_lock<std::mutex> lck(thread_para->mtx);
+      if (thread_para->index >= thread_para->wavs_file_name.size())
+        break;
+      
+      utt = thread_para->utts[thread_para->index];
+      wav_file_name = thread_para->wavs_file_name[thread_para->index];
+      thread_para->index++;
+    }
+
+
+    OnlineNnet2FeaturePipeline feature_pipeline(*(thread_para->pipeline_info));
+    DecodableNnetBatchLoopedOnline decodable(
+        thread_para->computer, *(thread_para->trans_model), 
+        feature_pipeline.InputFeature(), 
+        feature_pipeline.IvectorFeature());
+    std::vector< Vector<BaseFloat> > likelihoods;
+    
+    wav_file = fopen(wav_file_name.c_str(), "rb");
```

Comment:
We should avoid to use the C style IO, after all Kaldi is a C++ project.
`WaveData` in `src/feat/wave-reader.h` should be a better choice.

Reply:
See
https://github.com/kaldi-asr/kaldi/blob/220234184f17c002154ac36de2c79932afba112e/src/feat/wave-reader.h#L191

You can find example usages in
https://github.com/kaldi-asr/kaldi/blob/816de2ac76c5159fb6deb919c69a11d2f3434107/src/featbin/wav-copy.cc#L72-L75

Reply:
ok, I will fix it.